pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -41,46 +41,66 @@ const (
 	ManagementClusterHostname = "mgmtCluster"
 )
 
-// TODO: Need to do inheritance of DestRules based on domain suffix match
-
 // BuildClusters returns the list of clusters for the given proxy. This is the CDS output
 // For outbound: Cluster for each service/subset hostname or cidr with SNI set to service hostname
 // Cluster type based on resolution
 // For inbound (sidecar only): Cluster for each inbound endpoint port and for each service port
 func (configgen *ConfigGeneratorImpl) BuildClusters(env *model.Environment, proxy *model.Proxy, push *model.PushContext) ([]*v2.Cluster, error) {
 	clusters := make([]*v2.Cluster, 0)
 
-	recomputeOutboundClusters := true
-	if configgen.CanUsePrecomputedCDS(proxy) {
-		if configgen.PrecomputedOutboundClusters != nil && configgen.PrecomputedOutboundClusters[proxy.ConfigNamespace] != nil {
-			clusters = append(clusters, configgen.PrecomputedOutboundClusters[proxy.ConfigNamespace]...)
-			recomputeOutboundClusters = false
-		}
-	}
-
-	if recomputeOutboundClusters {
-		clusters = append(clusters, configgen.buildOutboundClusters(env, proxy, push)...)
+	instances, err := env.GetProxyServiceInstances(proxy)
+	if err != nil {
+		log.Errorf("failed to get service proxy service instances: %v", err)
+		return nil, err
 	}
+	clusters = append(clusters, configgen.buildOutboundClusters(env, push, proxy, instances)...)
 
-	if proxy.Type == model.Sidecar {
-		instances, err := env.GetProxyServiceInstances(proxy)
-		if err != nil {
-			log.Errorf("failed to get service proxy service instances: %v", err)
-			return nil, err
-		}
-
+	switch proxy.Type {
+	case model.SidecarProxy:
 		// Let ServiceDiscovery decide which IP and Port are used for management if
 		// there are multiple IPs
 		managementPorts := make([]*model.Port, 0)
 		for _, ip := range proxy.IPAddresses {
 			managementPorts = append(managementPorts, env.ManagementPorts(ip)...)
 		}
 		clusters = append(clusters, configgen.buildInboundClusters(env, proxy, push, instances, managementPorts)...)
+	case model.Router:
+		if proxy.GetRouterMode() == model.SniDnatRouter {
+			clusters = append(clusters, configgen.buildOutboundSniDnatClusters(env, proxy, push)...)
+		}
 	}
 
-	if proxy.Type == model.Router && proxy.GetRouterMode() == model.SniDnatRouter {
-		clusters = append(clusters, configgen.buildOutboundSniDnatClusters(env, proxy, push)...)
-	}
+	// recomputeOutboundClusters := true
+	// if configgen.CanUsePrecomputedCDS(proxy) {
```

Comment:
oh oh, this cache eventually lose effect.

Reply:
If we reduce the number of clusters sent to each sidecar - we may not need the cache, or we can have a per-proxy cache ( see above discussion on on-demand).